completions/make

```diff
@@ -1,91 +1,92 @@
-
-    # The stuff above here describes lines that are not
-    #  explicit targets or not targets other than special ones
-    # The stuff below here decides whether an explicit target
-    #  should be output.
-
-    /^# * File is an intermediate prerequisite/ {
-      s/^.*$//;x;                               # unhold target
-      d;                                        # delete line
-    }
-
-    /^$/ {                                      # end of target block
-      x;                                        # unhold target
-      /^$/d;                                    # dont print blanks
-      s|^${dirname_re-}\(.\{${#basename}\}[^:]*\):.*$|${output}|p;
-      d;                                        # hide any bugs
-    }
-
-    # This pattern includes a literal tab character as \t is not a portable
-    # representation and fails with BSD sed
-    /^[^#	:%]\{1,\}:/ {         # found target block
-      /^\.PHONY:/                 d;            # special target
-      /^\.SUFFIXES:/              d;            # special target
-      /^\.DEFAULT:/               d;            # special target
-      /^\.PRECIOUS:/              d;            # special target
-      /^\.INTERMEDIATE:/          d;            # special target
-      /^\.SECONDARY:/             d;            # special target
-      /^\.SECONDEXPANSION:/       d;            # special target
-      /^\.DELETE_ON_ERROR:/       d;            # special target
-      /^\.IGNORE:/                d;            # special target
-      /^\.LOW_RESOLUTION_TIME:/   d;            # special target
-      /^\.SILENT:/                d;            # special target
-      /^\.EXPORT_ALL_VARIABLES:/  d;            # special target
-      /^\.NOTPARALLEL:/           d;            # special target
-      /^\.ONESHELL:/              d;            # special target
-      /^\.POSIX:/                 d;            # special target
-      /^\.NOEXPORT:/              d;            # special target
-      /^\.MAKE:/                  d;            # special target
-EOF
-
-    # don't complete with hidden targets unless we are doing a partial completion
-    if [[ ! ${prefix_pat} || ${prefix_pat} == */ ]]; then
-        cat <<EOF
-      /^${prefix_pat}[^a-zA-Z0-9]/d;            # convention for hidden tgt
-EOF
-    fi
-
-    cat <<EOF
-      h;                                        # hold target
-      d;                                        # delete line
-    }
-
-EOF
+    local mode=$1
+    local -x prefix=$2
+
+    # display mode, only output current path component to the next slash
+    local -x prefix_replace=$prefix
+    [[ $mode == -d && $prefix == */* ]] &&
+        prefix_replace=${prefix##*/}
+
+    awk '
+      BEGIN {
+        prefix = ENVIRON["prefix"];
+        prefix_replace = ENVIRON["prefix_replace"];
+        is_target_block = 0;
+        target = "";
+      }
+      function starts_with(str, prefix) {
```

Comment:
Solaris awk lacks many features. It doesn't have a user-defined function as you guessed. It doesn't support `ENVIRON` It doesn't support `-v var=value` command-line arguments. It doesn't seem to support the condition of the form `!expr` in the block beginning.

Reply:
One way is to use `/usr/xpg4/bin/awk`. Solaris intensionally preserves the ancient feature set of core utilities, but at the same time they provide POSIX utilities in `/usr/xpg4/bin`. Another way is to use `awk` with a limited set of features, which I'm not even sure if it is possible.

Reply:
Maybe trying to limp along with the simplest of awk would lose the readability benefits of using awk in the first place.

Reply:
I separated the AWK script in commit db3a82d46c67fb6320967158e69c0022a251b3e3.

For Solaris awk, I found also other cases where the feature that Solaris awk does not support is used. In commit 6a9b7fe41f76fe863cd07f36d48b0c5e48f6c7e4, I added a utility `_comp_awk` (which calls `/usr/xpg4/bin/awk "$@"` in Solaris or otherwise `awk "$@"`) and replaced a few calls of `awk` that need POSIX features. The use of `ENVIRON` in `portinstall` was recently introduced by me in #1053.

In commit 36a2e375c, I have modified the condition `$NF=2` to `NF >= 2`. Solaris awk doesn't like the assignment `$NF=2` in the block condition, but to begin with, I doubt this condition. See the commit message for the details.